src/SoftWire.cpp

```diff
@@ -4,8 +4,58 @@
 // protection against spurious signals on the I2C bus is only available
 // for AVR architectures where ATOMIC_BLOCK is defined.
 
+//
+// 2022-04-13 by Technik Gegg
+//
+//    Added a workaround for the missing ATOMIC_BLOCK in non AVR architectures, 
+//    which I've found on Stackoverflow (see link in comments below).
+//
+
 #if defined(ARDUINO_ARCH_AVR)
 #include <util/atomic.h>
+#define CRITICAL_SECTION ATOMIC_BLOCK(ATOMIC_RESTORESTATE)
+
+#elif defined(ARDUINO_ARCH_SAM)
+  // Workaround as suggested by Stackoverflow user "Notlikethat"
+  // http://stackoverflow.com/questions/27998059/atomic-block-for-reading-vs-arm-systicks
+
+  static inline int __int_disable_irq(void) {
+    int primask;
+    asm volatile("mrs %0, PRIMASK\n" : "=r"(primask));
+    asm volatile("cpsid i\n");
+    return primask & 1;
+  }
+
+  static inline void __int_restore_irq(int *primask) {
+    if (!(*primask)) {
+      asm volatile ("" ::: "memory");
+      asm volatile("cpsie i\n");
+    }
+  }
+  // This critical section macro borrows heavily from
+  // avr-libc util/atomic.h
+  // --> http://www.nongnu.org/avr-libc/user-manual/atomic_8h_source.html
+  #define CRITICAL_SECTION for (int primask_save __attribute__((__cleanup__(__int_restore_irq))) = __int_disable_irq(), __ToDo = 1; __ToDo; __ToDo = 0)
+
+#elif defined(ARDUINO_ARCH_STM32)
+#include <Arduino.h>
+#include <cmsis_gcc.h>
+	// exact same as above only using predefined CMSIS functions
+	
+  static inline int __int_disable_irq(void) {
+    int primask = __get_PRIMASK();
+    __disable_irq();
+    return primask & 1;
+  }
+
+  static inline void __int_restore_irq(int *primask) {
+    if (!(*primask)) {
+      __enable_irq();
+    }
+  }
+  #define CRITICAL_SECTION for (int primask_save __attribute__((__cleanup__(__int_restore_irq))) = __int_disable_irq(), __ToDo = 1; __ToDo; __ToDo = 0)
+#else
+  #error Unsupported controller architecture
```

Comment:
Doesn't this completely break the library for non-AVR and non-STM32 users, like ESP32? Issuing a pragma message seems more in keeping with the current library behaviour.

Reply:
What would you suggest?
Assigning an empty #define CRITICAL_SECTION and adding a warning instead to let the user know there might be a problem?

Reply:
How about the conditional logic part starts with 

~~~
#ifdef CRITICAL_SECTION
// User has created their own workaround. Do nothing more here.

#elif 

// Your CRITICAL_SECTION stuff goes here

#else
  #pragma message "Your controller architecture is missing interrupt protection"
#endif
~~~

The first part means users can somehow add their own protection method externally to SoftWire. The last part warns users there is a problem. In production code I often have the compiler treat warnings as errors. However a library user cannot change their code to fix the library so I'd prefer a pragma message here. I realise that a `#define #CRITICAL_SECTION` could be used to silence the program/warning but the Arduino IDE doesn't provide a method for users to define macros for all compilation units, and I gave up fighting that battle long ago.

Reply:
Sounds like a feasible solution to me.